### ISSUE TITLE
compile errors when building with python3 support

### ISSUE THREAD

**Issue by micskeiz** [phatina/python-lmiwbem#21]
I tried to build the project with python3 support, but the build failed.

Environment:
* Setup: CentOS 7 minimal 64 bit, python3.3 (from softwarecollections.org), python-lmiwbem source 0.6.1-pre.
* Build process: as outlined in the readme (autogen, configure, make, make install)
* Configure parameters: `../configure --with-listener=no --with-slp=no --with-python3=yes`

I encountered the following errors.

### lmiwbem_property.cpp error

The make phase produced the following error:

```
../../src/obj/cim/lmiwbem_property.cpp: In member function 'bool CIMProperty::eq(const boost::python::api::object&)':
../../src/obj/cim/lmiwbem_property.cpp:257:26: error: 'getValue' was not declared in this scope
         compare(getValue(), cim_other.getValue(), Py_EQ) &&
                          ^
../../src/obj/cim/lmiwbem_property.cpp:257:39: error: 'class CIMProperty' has no member named 'getValue'
         compare(getValue(), cim_other.getValue(), Py_EQ) &&
                                       ^
../../src/obj/cim/lmiwbem_property.cpp: In member function 'bool CIMProperty::gt(const boost::python::api::object&)':
../../src/obj/cim/lmiwbem_property.cpp:275:26: error: 'getValue' was not declared in this scope
         compare(getValue(), cim_other.getValue(), Py_GT) ||
                          ^
../../src/obj/cim/lmiwbem_property.cpp:275:39: error: 'class CIMProperty' has no member named 'getValue'
         compare(getValue(), cim_other.getValue(), Py_GT) ||
                                       ^
../../src/obj/cim/lmiwbem_property.cpp: In member function 'bool CIMProperty::lt(const boost::python::api::object&)':
../../src/obj/cim/lmiwbem_property.cpp:293:26: error: 'getValue' was not declared in this scope
         compare(getValue(), cim_other.getValue(), Py_LT) ||
                          ^
../../src/obj/cim/lmiwbem_property.cpp:293:39: error: 'class CIMProperty' has no member named 'getValue'
         compare(getValue(), cim_other.getValue(), Py_LT) ||
```

The problem is that in lmiwbem_property.cpp the methods `CIMProperty::eq`, `CIMProperty::gt`, `CIMProperty::lt` are only built for Python3, and they contain similar lines:

    compare(getValue(), cim_other.getValue(), Py_EQ) && 

getValue() is not defined. If I change them to getPyValue(), then the code compiles (but I do not know the code of lmiwbem enough to see whether this is the intended behavior).

### reference to PyObject_Unicode

The next error is the following:

```
../../src/util/lmiwbem_convert.cpp: In static member function 'static boost::python::api::object ObjectConv::asPyUnicode(const boost::python::api::object&)':
../../src/util/lmiwbem_convert.cpp:340:62: error: 'PyObject_Unicode' was not declared in this scope
     return bp::object(bp::handle<>(PyObject_Unicode(obj.ptr())));
```

In Python 3.x as far as I can see `PyObject_Unicode` was removed (see https://docs.python.org/3.4/c-api/object.html).

After commenting out the ObjectConv::asPyUnicode method, compiling finishes.

### linker error 

Next is a linker error:

```
/bin/ld: cannot find -l-L/usr/lib/gcc/x86_64-redhat-linux/4.8.2
collect2: error: ld returned 1 exit status
make[2]: *** [lmiwbem_core.la] Error 1
make[2]: Leaving directory `/root/lmiwbem-master/build/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/root/lmiwbem-master/build'
make: *** [all] Error 2
```

The problem is that the generated makefile (build/src/makefile) contains the following line:

```
lmiwbem_core_la_LIBADD = -lpegcommon -lpegclient \
        -L/opt/rh/python33/root/usr/lib64 -lpython3.3m -l $(am__append_4) \
        $(am__append_5) $(am__append_8)
```

The extra `-l` is not needed there. I could not trace back why this empty `-l` is generated there, probably a module was not found in my setup.

### setup.py.in error

I tried also `python setup.py install`, but in setup.py.in there is a print using python 2.x syntax. It could be changed to print():

```
diff --git a/setup.py.in b/setup.py.in
index baef49b..ad74d0e 100644
--- a/setup.py.in
+++ b/setup.py.in
@@ -88,7 +88,7 @@ class lmiwbem_build(build):
 
         # Generate pydoc headers
         for pydoc, header in zip(pydocs, headers):
-            print 'docgen %s %s' % (pydoc, header)
+            print('docgen %s %s' % (pydoc, header))
             subprocess.check_call([lmiwbem_docgen, pydoc, header])
 
     def teardown(self):
```

### boost_python3

It could be noted in the readme, that if building for python3, then the boost_python3 version of boost_python is needed (it is not trivial in some distributions).

**Comment by phatina** [comment]
I will make the process of compiling with Python3 more clear. Thank you!

**Comment by phatina** [comment]
Please, check the current master if --with-python3 works for you (commit aee03f8).

Wrt makefile: running Fedora 21, I see "-lboost_python3" right after "-lpython3.4m".  Do you have boost_python listed in LIBADD list?

**Comment by micskeiz** [comment]
*Compile*: Thanks, with the new commit now it compiles!

*Regarding the `-l` problem*: yes, the problem is boost_python3.

In my setup I had the boost_python lib installed, but not boost_python3. In this case configure detects that I have BOOST_PYTHON installed, but does not set the library name.

Relevant part from configure-s output:

```
checking for boostlib >= 1.40... yes
checking whether the Boost::Python library is available... yes
checking for exit in -lboost_python3... no
```

From config.status:

```
S["BOOST_PYTHON_LIB"]="-l"
...
S["PYTHON_VERSION"]="3.3"
...
D["HAVE_PYTHON"]=" \"3.3\""
D["HAVE_BOOST"]=" /**/"
D["HAVE_BOOST_PYTHON"]=" /**/"
```

As far as I can see boost_python.m4 checks first for the `boost_python` lib and sets `HAVE_BOOST_PYTHON`. Then it checks the version and tries to find the lib name for `boost_python3`, but it could not find it, thus clears it.

I checked with version 20 of ax_boost_python (from: http://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_boost_python.m4), but the result was the same.

**Comment by phatina** [comment]
Fine, I will try to play with boost_python3. Once I figure this out, I can release 0.7.1...

**Comment by phatina** [comment]
Try this:

``` diff
diff --git a/configure.ac b/configure.ac
index c93d417..e8e63d7 100644
--- a/configure.ac
+++ b/configure.ac
@@ -225,8 +225,12 @@ dnl Check for Boost::Python
 AX_BOOST_BASE([1.40], [],
     [AC_MSG_ERROR([Boost not found; install boost-devel])])
 AX_BOOST_PYTHON([$PYTHON_VERSION])
-if test x"$ac_cv_boost_python" = x"no"; then
-    AC_MSG_ERROR(Boost::Python not found; install boost-python)
+if test x"$BOOST_PYTHON_LIB" = x""; then
+    if test x"${PYTHON_VERSION:0:1}" = x"3"; then
+        AC_MSG_ERROR(Boost::Python3 not found; install boost-python3)
+    else
+        AC_MSG_ERROR(Boost::Python not found; install boost-python)
+    fi
 fi
 
 dnl Note that AX_BOOST_PYTHON just gives us the name of the library (BOOST_PYTHON_LIB),
```

**Comment by micskeiz** [comment]
Works perfectly in my setup. Thanks for the quick help and fixes.

**Comment by phatina** [comment]
Fine. Thank you for the report.

**Comment by phatina** [comment]
Fixed in 931db8d.
